tests/store/test_sqlalchemy_store.py

```diff
@@ -498,3 +507,173 @@ def test_restore_run(self):
         restored = self.store.get_run(run_uuid)
         self.assertEqual(restored.info.run_uuid, run_uuid)
         self.assertEqual(restored.info.lifecycle_stage, entities.LifecycleStage.ACTIVE)
+
+    # Tests for Search API
+    def _search(self, experiment_id, metrics_expressions=None, param_expressions=None,
+                run_view_type=ViewType.ALL):
+        conditions = (metrics_expressions or []) + (param_expressions or [])
+        return [r.info.run_uuid
+                for r in self.store.search_runs([experiment_id], conditions, run_view_type)]
+
+    def _param_expression(self, key, comparator, val):
+        expr = SearchExpression()
+        expr.parameter.key = key
+        expr.parameter.string.comparator = comparator
+        expr.parameter.string.value = val
+        return expr
+
+    def _metric_expression(self, key, comparator, val):
+        expr = SearchExpression()
+        expr.metric.key = key
+        expr.metric.double.comparator = comparator
+        expr.metric.double.value = val
+        return expr
+
+    def test_search_vanilla(self):
+        exp = self._experiment_factory('search_vanilla').experiment_id
+        runs = [self._run_factory('r_%d' % r, exp).run_uuid for r in range(3)]
+
+        self.assertSequenceEqual(runs, self._search(exp, run_view_type=ViewType.ALL))
+        self.assertSequenceEqual(runs, self._search(exp, run_view_type=ViewType.ACTIVE_ONLY))
+        self.assertSequenceEqual([], self._search(exp, run_view_type=ViewType.DELETED_ONLY))
+
+        first = runs[0]
+
+        self.store.delete_run(first)
+        self.assertSequenceEqual(runs, self._search(exp, run_view_type=ViewType.ALL))
+        self.assertSequenceEqual(runs[1:], self._search(exp, run_view_type=ViewType.ACTIVE_ONLY))
+        self.assertSequenceEqual([first], self._search(exp, run_view_type=ViewType.DELETED_ONLY))
+
+        self.store.restore_run(first)
+        self.assertSequenceEqual(runs, self._search(exp, run_view_type=ViewType.ALL))
+        self.assertSequenceEqual(runs, self._search(exp, run_view_type=ViewType.ACTIVE_ONLY))
+        self.assertSequenceEqual([], self._search(exp, run_view_type=ViewType.DELETED_ONLY))
+
+    def test_search_params(self):
+        experiment_id = self._experiment_factory('search_params').experiment_id
+        r1 = self._run_factory('r1', experiment_id).to_mlflow_entity().info.run_uuid
+        r2 = self._run_factory('r2', experiment_id).to_mlflow_entity().info.run_uuid
+
+        self.store.log_param(r1, entities.Param('generic_param', 'p_val'))
+        self.store.log_param(r2, entities.Param('generic_param', 'p_val'))
```

Comment:
nit: in order to avoid potential mix-up of these two parameters, maybe we could give them different values for the different runs.

Reply:
oh, I see that it's important for the search query to have the same value. Since we have two pairs of parameters:
- same key and same value
- different keys and different values

Maybe it would be good to add "same key and different values" as a third test case here.

Reply:
aah! no -- i wanted test that search returns both runs